File: app/src/main/webapp/js/validator.js
// Check if a <= b
function compare(a, b){
    if (a.toString()[0] == "-"){
        if (b.toString()[0] == "-"){
            return compare(b.toString().slice(1), a.toString().slice(1));
        }
        return true;
    }

    let ax = (a.toString() + ".0").split('.');
    let bx = (b.toString() + ".0").split('.');
    let maxLen = Math.max(ax[1].length, bx[1].length);
    ax[1] = ax[1] + '0'.repeat(maxLen - ax[1].length);
    bx[1] = bx[1] + '0'.repeat(maxLen - bx[1].length);
    if (ax[0] !== bx[0]){
        return parseFloat(ax[0]) <= parseFloat(bx[0]);
    }
    for (let i = 0; i < maxLen; i++) {
        if (ax[1][i] !== bx[1][i])
            return ax[1][i] < bx[1][i];
    }
    return true;
}

export class Validator {
    static validate(x, y, r, drawn) {
        return (this.validateX(x) && this.validateY(y) && this.validateR(r)) ||
            (drawn && this.validateR(r));
    }

    static validateX(x){
        return ["-2", "-1.5", "-1", "-0.5", "0", "0.5", "1", "1.5", "2"].includes(x)
    }

    static validateY(y){
        return compare(-3, y) && compare(y, 5);
    }

    static validateR(r){
        return compare(1, r) && compare(r, 4);
    }

}

File: app/src/main/java/app/utils/Validator.java
package app.utils;

import java.util.ArrayList;

import app.models.Point;

public class Validator {
    private ArrayList<Float> x_values = new ArrayList<Float>(java.util.Arrays.asList(-2f, -1.5f, -1f, -0.5f, 0f, 0.5f, 1f, 1.5f, 2f));
    private float min_y = -3f;
    private float max_y = 5f;
    private float min_r = 1f;
    private float max_r = 4f;
    public Point validate(String x, String y, String r, String drawn) {
        Point point = new Point();
        try{
            point.setX(Float.parseFloat(x));
            point.setY(Float.parseFloat(y));
            point.setR(Float.parseFloat(r));
        } catch (Exception e) {
            point.setValid(false);
            return point;
        }
        point.setValid(validate_data(
            point.getX(),
            point.getY(),
            point.getR()
        ));
        try{
           point.setDrawn(Boolean.parseBoolean(drawn));
        } catch (Exception e) {
            point.setDrawn(false);
        }
        point.setValid(point.isValid() || (point.isDrawn() && validate_r(point.getR())));

        return point;
    }

    private boolean validate_data(float x, float y, float r) {
        return validate_x(x) && validate_y(y) && validate_r(r);
    }

    private boolean validate_x(float x) {
        return x_values.contains(x);
    }

    private boolean validate_y(float y) {
        return min_y <= y && y <= max_y;
    }

    private boolean validate_r(float r) {
        return min_r <= r && r <= max_r;
    }
}
